src/mappings/utils/cumulativeVolumes.ts

```diff
@@ -228,73 +233,357 @@ async function createPooledAmount(swapAmount: SwapDetailsAmount): Promise<Pooled
 /**
  * Modifies the passed in entity to add or update pooled token volumes.
  * @param entity The entity to add pooled amounts to, or update existing ones
- * @param swapDetails The swap details used to modify the event
+ * @param amounts The swap detail amounts to add to the entity's values
  * @return Returns the modified entity
  */
-async function updateOrAddPooledAmounts(entity: CumulativeDexTradingVolumePerPool, swapDetails: SwapDetails): Promise<CumulativeDexTradingVolumePerPool> {
-    // we need to find & update existing amounts, or add new ones
-    let foundFrom = false;
-    let foundTo = false;
-    
-    for (const pooledAmount of entity.amounts) {
-        if (foundFrom && foundTo) {
-            // done: exit loop
-            break;
+async function increaseSwapDetailsAmountsForEntity<T extends CumulativeDexTradingVolume | CumulativeDexTradingVolumePerAccount>(
+    entity: T,
+    amounts: SwapDetailsAmount[]
+): Promise<T> {
+    for (const amountToAdd of amounts) {
+        let updatedExistingAmount = false;
+        for (const pooledAmount of entity.amounts) {
+            if (isEqual(pooledAmount.token, amountToAdd.currency)) {
+                const newAmount = pooledAmount.amount + amountToAdd.atomicAmount;
+                pooledAmount.amount = newAmount;
+                pooledAmount.amountHuman = await convertAmountToHuman(pooledAmount.token, newAmount);
+                updatedExistingAmount = true;
+                break;
+            }
         }
```

Comment:
Is this correct way to compute the volume? I am not entirely sure, but won't this cause the total volume to be 2 times bigger than it actually is? 

Let's say that there is 100USD -> BTC trade, then 100usd will be added to total cumulative volumes, but also BTC equivalent of 100USD that will be traded because both input and output amounts are included. Therefore the total volume is not 100USD, but ~200 USD now.

Reply:
> Is this correct way to compute the volume? I am not entirely sure, but won't this cause the total volume to be 2 times bigger than it actually is?

Yes, great question!
Squid will indeed count both, but it also stores it in the traded currencies (without conversions). So, with your example:
A trade of 100 USDT -> 0.004 BTC (round numbers), then it will add two entries to the pooled amounts (format simplified for legibility):
```
[ 
  { currency: USDT, amountHuman: 100, ... },
  { currency: BTC: amountHuman: 0.004, ... }
]
```
Then, when another trade is done, let's say the other way around: 0.002 BTC -> 50 USDT, the pooled amounts get updated to:
```
[
  { currency: USDT, amountHuman: 150, ... },
  { currency: BTC, amountHuman: 0.006, ... }
]
```

Reply:
I think this is also a good point to raise with @nud3l (👋 ) , just to make sure if we want that bahavior for trade volume counts

Reply:
Counting both sides of the trade is correct and in line with how Uniswap v2 calculates trading volume as well: https://github.com/Uniswap/v2-subgraph/blob/master/src/mappings/core.ts#L399

I can recommend that subgraph as a reference btw. I look there usually when I'm not sure :)